pkg/relay/ethereum/tx.go

```diff
@@ -504,107 +518,105 @@ func (iter *CallIter) Current() sdk.Msg {
+func (iter *CallIter) Next(n int) {
+	iter.cursor = min(len(iter.msgs), iter.cursor + n)
+}
+
+func (iter *CallIter) updateLoggerMessageInfo(logger *log.RelayLogger, from int, count int) {
+	if from < 0 || count < 0 || len(iter.msgs) <= from || len(iter.msgs) < from + count {
+		logger.Error("invalid parameter", fmt.Errorf("out of index: len(msgs)=%d, from=%d, count=%d", len(iter.msgs), from, count))
+		return
 	}
+
+	logger.Logger = logger.With(
+		logAttrMsgIndexFrom, from,
+		logAttrMsgCount, count,
+		logAttrMsgType, strings.Join(iter.msgTypeNames[0 : from + count], ","),
+	)
 }
 
-func (iter *CallIter) SendTx(ctx context.Context, c *Chain) (*gethtypes.Transaction, error) {
+func (iter *CallIter) BuildTx(ctx context.Context, c *Chain) (*gethtypes.Transaction, int, error) {
 	if c.multicall3 == nil {
-		return iter.sendSingleTx(ctx, c)
+		return iter.buildSingleTx(ctx, c)
 	} else {
-		return iter.sendMultiTx(ctx, c)
+		return iter.buildMultiTx(ctx, c)
 	}
 }
 
-func (iter *CallIter) sendSingleTx(ctx context.Context, c *Chain) (*gethtypes.Transaction, error) {
+func (iter *CallIter) buildSingleTx(ctx context.Context, c *Chain) (*gethtypes.Transaction, int, error) {
 	if iter.End() {
-		return nil, nil
+		return nil, 0, nil
 	}
 
 	logger := c.GetChainLogger()
-	logger = &log.RelayLogger{Logger: logger.With(
-		logAttrMsgIndexFrom, iter.Cursor(),
-		logAttrMsgType, fmt.Sprintf("%T", iter.Current()),
-	)}
+	iter.updateLoggerMessageInfo(logger, iter.Cursor(), 1)
 
 	opts, err := c.TxOpts(ctx, true);
 	if err != nil {
-		return nil, err
+		return nil, 0, err
 	}
 
 	// gas estimation
 	{
 		opts.GasLimit = math.MaxUint64
 		opts.NoSend = true
-		tx, err := c.SendTx(opts, iter.Current(), iter.skipUpdateClientCommitment)
+		tx, err := c.BuildMessageTx(opts, iter.Current(), iter.skipUpdateClientCommitment)
 		if err != nil {
 			logger.Error("failed to build tx for gas estimation", err)
-			return nil, err
+			return nil, 1, err
 		}
 
 		txGasLimit, err := estimateGas(ctx, c, tx, true, logger)
 		if err != nil {
-			return nil, err
+			return nil, 1, err
 		}
 		opts.GasLimit = txGasLimit
 	}
 
-	opts.NoSend = false
-	tx, err := c.SendTx(opts, iter.Current(), iter.skipUpdateClientCommitment)
+	tx, err := c.BuildMessageTx(opts, iter.Current(), iter.skipUpdateClientCommitment)
 	if err != nil {
-		logger.Error("failed to send msg", err)
-		return nil, err
+		logger.Error("failed to build tx", err)
+		return nil, 0, err
 	}
-	iter.Next()
-	return tx, nil
+
+	return tx, 1, nil
 }
 
-func (iter *CallIter) sendMultiTx(ctx context.Context, c *Chain) (*gethtypes.Transaction, error) {
+func (iter *CallIter) buildMultiTx(ctx context.Context, c *Chain) (*gethtypes.Transaction, int, error) {
 	if (iter.End()) {
-		return nil, nil
+		return nil, 0, nil
 	}
 	// now iter.cursor < len(iter.msgs)
 
 	logger := c.GetChainLogger()
 
 	opts, err := c.TxOpts(ctx, true);
 	if err != nil {
-		if err != nil {
-			return nil, err
-		}
+		return nil, 1, err
```

Comment:
The second return value of `buildMultiTx` sometimes becomes `0` and at other times takes on different values.
I couldn't understand the criteria for that.

Reply:
same as above.